drivers/counter/counter_neorv32_gptmr.c

```diff
@@ -0,0 +1,278 @@
+
+	ctrl = neorv32_gptmr_read(dev, NEORV32_GPTMR_CTRL);
+	ctrl |= NEORV32_GPTMR_CTRL_EN;
+	neorv32_gptmr_write(dev, NEORV32_GPTMR_CTRL, ctrl);
+
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+
+static int neorv32_gptmr_stop(const struct device *dev)
+{
+	struct neorv32_gptmr_data *data = dev->data;
+	k_spinlock_key_t key;
+	uint32_t ctrl;
+
+	key = k_spin_lock(&data->lock);
+
+	ctrl = neorv32_gptmr_read(dev, NEORV32_GPTMR_CTRL);
+	ctrl &= ~NEORV32_GPTMR_CTRL_EN;
+	neorv32_gptmr_write(dev, NEORV32_GPTMR_CTRL, ctrl);
+
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+
+static int neorv32_gptmr_get_value(const struct device *dev, uint32_t *ticks)
+{
+	*ticks = neorv32_gptmr_read(dev, NEORV32_GPTMR_COUNT);
+
+	return 0;
+}
+
+static int neorv32_gptmr_set_top_value(const struct device *dev, const struct counter_top_cfg *cfg)
+{
+	static const uint32_t supported =
+		COUNTER_TOP_CFG_DONT_RESET | COUNTER_TOP_CFG_RESET_WHEN_LATE;
+	struct neorv32_gptmr_data *data = dev->data;
+	k_spinlock_key_t key;
+	bool restart = false;
+	uint32_t count;
+	uint32_t ctrl;
+	int err = 0;
+
+	__ASSERT_NO_MSG(cfg != NULL);
+
+	if (cfg->ticks == 0) {
+		return -EINVAL;
+	}
+
+	if ((cfg->flags & ~supported) != 0U) {
+		LOG_ERR("unsupported flags 0x%08x", cfg->flags);
+		return -ENOTSUP;
+	}
+
+	data->top_callback = cfg->callback;
+	data->top_user_data = cfg->user_data;
+
+	key = k_spin_lock(&data->lock);
+
+	ctrl = neorv32_gptmr_read(dev, NEORV32_GPTMR_CTRL);
+	count = neorv32_gptmr_read(dev, NEORV32_GPTMR_COUNT);
+
+	if ((ctrl & NEORV32_GPTMR_CTRL_EN) != 0U) {
+		if ((cfg->flags & COUNTER_TOP_CFG_DONT_RESET) == 0U) {
+			neorv32_gptmr_write(dev, NEORV32_GPTMR_CTRL, ctrl & ~NEORV32_GPTMR_CTRL_EN);
+			restart = true;
+		} else if (count >= cfg->ticks) {
+			if ((cfg->flags & COUNTER_TOP_CFG_RESET_WHEN_LATE) != 0U) {
+				neorv32_gptmr_write(dev, NEORV32_GPTMR_CTRL,
+						    ctrl & ~NEORV32_GPTMR_CTRL_EN);
+				restart = true;
+			}
+
+			err = -ETIME;
+		}
+	}
+
+	neorv32_gptmr_write(dev, NEORV32_GPTMR_THRES, cfg->ticks);
+
+	if (restart) {
+		neorv32_gptmr_write(dev, NEORV32_GPTMR_CTRL, ctrl);
+	}
+
+	k_spin_unlock(&data->lock, key);
+
+	return err;
+}
+
+static uint32_t neorv32_gptmr_get_pending_int(const struct device *dev)
+{
+	uint32_t ctrl;
+
+	ctrl = neorv32_gptmr_read(dev, NEORV32_GPTMR_CTRL);
+
+	return (ctrl & NEORV32_GPTMR_CTRL_IRQ_PND) != 0 ? 1 : 0;
+}
+
+static uint32_t neorv32_gptmr_get_top_value(const struct device *dev)
+{
+	return neorv32_gptmr_read(dev, NEORV32_GPTMR_THRES);
+}
+
+static uint32_t neorv32_gptmr_get_freq(const struct device *dev)
+{
+	static const uint16_t prescalers[] = { 2U, 4U, 8U, 64U, 128U, 1024U, 2048U, 4096U };
+	const struct neorv32_gptmr_config *config = dev->config;
+	uint32_t clk;
+	int err;
+
+	err = syscon_read_reg(config->syscon, NEORV32_SYSINFO_CLK, &clk);
+	if (err < 0) {
+		LOG_ERR("failed to determine clock rate (err %d)", err);
+		return 0U;
+	}
+
+	return clk / prescalers[config->prescaler];
```

Comment:
optional: if you turn `prescalers` into an array of shift counts, then the prescaler list can be a uint8_t and this divide can become a shift.

This will slightly reduce the flash footprint.

Reply:
The compiler is smart enough to optimise this on its own. The suggested solution using a bitshift look-up table yields the exact same flash use as the current implementation.

Reply:
Keeping this for readability's sake is fine by me. I ran the variants through godbolt and couldn't get the compiler to optimise the divide into a shift: https://godbolt.org/z/sxhMe6ne1